### ISSUE TITLE
Ability to select different currency for each transaction

### ISSUE THREAD

**Issue by vladartym** [spliit-app/spliit#53]
FEATURE REQUEST: Ability to select different currency for each transaction, similar to splitwise. And choose the currency for reimbursment

**Comment by scastiel** [comment]
Interesting suggestion! So that would mean:
- defining a main currency for the group (among known currencies we are able to convert)
- selecting the currency for each expense (defaulting to the group one)
- calculating the converted amount for each expense (maybe storing it in the expense as well), ideally with the rate at the date of the expense, but maybe easier with the rate at the time the expense is entered.

We’d need to call an API to get conversion rates between currencies. Any API suggestion (cheap, or even better, free) is appreciated. Note that there wouldn’t be that many API calls, if we cache the conversion rates properly.

**Comment by vladartym** [comment]
> We’d need to call an API to get conversion rates between currencies. Any API suggestion (cheap, or even better, free) is appreciated. Note that there wouldn’t be that many API calls, if we cache the conversion rates properly.

Yep! Everything you mentioned is spot on! I just stumbled upon this guy: https://github.com/hakanensari/frankfurter
Gets updated everyday, I dont think we need anything more than that. https://api.frankfurter.app/latest

We can also get currency conversions based on the date of the transaction e.g. https://api.frankfurter.app/2019-01-04

**Comment by mertd** [comment]
What do you think about displaying the rate in an input? Allowing to overwrite the live rate may make the app more resilient to service interruptions of the currency API (see the [recent issue at ihatemoney](https://github.com/spiral-project/ihatemoney/issues/1232)).

**Comment by vladartym** [comment]
> What do you think about displaying the rate in an input? Allowing to overwrite the live rate may make the app more resilient to service interruptions of the currency API (see the [recent issue at ihatemoney](https://github.com/spiral-project/ihatemoney/issues/1232)).

Great idea!!

**Comment by dmlls** [comment]
> We’d need to call an API to get conversion rates between currencies. Any API suggestion (cheap, or even better, free) is appreciated.

Might be worth taking a look at https://github.com/sal0max/currencies#features for further options and perhaps implementation details.

**Comment by ashaw93** [comment]
Just to chime in, maybe instead of doing the conversion, if there are outstanding balances in other currencies, show them as another section in the balance tab. That way you can keep adding custom currencies and not have to deal with exchanges.

**Comment by Big-Hammer** [comment]
I have a use case for this - I do a few classic car trips with a crew of 2 or 2 people. We often travel around Europe and incur costs in GBP, EUR and other currencies. Converting to GBP at the time of the expense with one call to the conversion service would be good - after all, we're not dealing in huge sums and we only ever settle up in our 'base' currency having incurred the costs many days or even weeks before, often in cash exchanged in the distant past (i.e. no-one's really tracking gentle fluctuations). This is probably the most time consuming element of settling up as in the past I've just sat there and manually converted it all with a nominal 'average' rate.

**Comment by NotWoods** [comment]
It would be better to just store the amount and currency in the database, then look up the historical exchange rate and calculate the converted amount later. (Other apps just look at the current exchange rate and calculate all at once.) If you want to support offline mode later the app shouldn't rely on an external API to write to the database.

**Comment by J0B10** [comment]
Has there been any progress on currency support?
I'm planning an USA-Trip with a couple of friends and we will have to mange expenses in both $ and €, so this would be a killer feature 😅.

**Comment by HuntClauss** [comment]
As for the exchange rate of a currency - in the case of currency exchange, the exchange office charges a certain percentage for itself, and thus the API data will be understated, the same for paying with a card in foreign currency. The best solution would be to get rate from API as a default with the possibility to change it manually to own rate

**Comment by niklas-uhl** [comment]
> The best solution would be to get rate from API as a default with the possibility to change it manually to own rate

If I remember correctly, this is also how Splitwise does it.

**Comment by delta0815** [comment]
I'm also waiting for this feature... it's in fact the feature that all open source solutions currently lacking. That's why I can't wait to see it and switch to spliit right away :-)

**Comment by Itschotsch** [comment]
Since this feature request has been open for almost one and a half years, I was wondering what is currently hindering development. Wouldn't @ashaw93's proposed solution be rather trivial to implement? At least as a working solution, independent from @whimcomp's proposed PR https://github.com/spliit-app/spliit/pull/330 until it is reviewed and merged. Let me know if I can help.

**Comment by wookey** [comment]
Another vote for this feature. I just used Spliit for a trip for the first time and it was great(!) but I was surprised to find it was a single-currency system (I'd expect Canadians to have to deal with USD on trips quite often?). Few of my use-cases are single-currency so this is an important feature (in fact the only major missing one). The implementation discussions above seem sound.

A few observations from experience: 

- One good reason for ability to specify a custom rate is credit-card payments which often include extra fees that change the effective rate.

- Currencies can in fact be anything - they are nothing but a name (normally currency symbol) and a conversions rate (optionally variable over time). So accounting is sometimes made simpler if you have a currency of 'nights' or 'beers' so you can just put in an expense of 30 nights, split into 7,7,7 and 9 'nights', and easily allocate a beer tally to people. Having both a long name (normally 'GBP', or 'USD' (but maybe 'beers') and a short name normally '£' or '$' is useful with this flexibility.

- The other two bits of software I use do it this way 'misery' https://packages.debian.org/sid/misery and 'BankofCUCC':  https://expo.survex.com/handbook/bankofexpo.html (for @delta0815 above - there _are_ open-source multi-currency solutions out there, but misery is local-only (_very_ cool to use, but local-only), and BoC/BoE was designed before the days of mobile phone screens, and needs a bit of behind-the-scenes setup for each instance, so is not as convenient as something like Spliit).

- Choosing the splits currency separately from the default currency is definitely useful. Our last trip was mostly UK people in Europe, so most expenses in Euro (default) but most settling-up payments in GBP. In this case displaying plsits in Euro and GBP would be most useful, so people can pay/be paid in whichever is easiest.

I presume the slow development is a matter of available time? Would you just like someone to get stuck in and have a go at an initial implementation?

**Comment by Petersmit27** [comment]
I'll check out #329 and #330 and most likely merge those after reviewing them. 

> Currencies can in fact be anything - they are nothing but a name (normally currency symbol) and a conversions rate (optionally variable over time). So accounting is sometimes made simpler if you have a currency of 'nights' or 'beers' so you can just put in an expense of 30 nights, split into 7,7,7 and 9 'nights', and easily allocate a beer tally to people. Having both a long name (normally 'GBP', or 'USD' (but maybe 'beers') and a short name normally '£' or '$' is useful with this flexibility.

I do like this idea, though I'll have to see if/how this would be possible with the implementation in the existing PRs.

**Comment by Petersmit27** [comment]
Hey there, it's been a bit but #330 has been merged. As a followup issue I created #412. Feel free to discuss possible approaches there!
